### PR TITLE
Add BRS for enforcing secure TLS version in Azure IoT Hub

### PR DESCRIPTION
**Pull request by rvadeghar**
(Azure_IoTHub_DP_Use_Secure_TLS_Version)

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 [hunk 1]
```diff
@@ -0,0 +1,878 @@
...
+            if(!$_.MinimumTlsVersion){
+                $TLSwithDefaultValue += $IoTHub
+            }
+        }
+        else
+        {
+            $logResource = @{}
+            $logResource.Add("ResourceGroupName",($_.ResourceGroupName))
+            $logResource.Add("ResourceName",($_.ResourceName))
+            $logResource.Add("Reason","Minimum required TLS Version configured set on IoT Hub.")
```
Comment (Aboli-msft) [inline Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:384]: Minimum required TLS Version configured set on IoT Hub-->Minimum required TLS Version is configured on IoT Hub

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+    {
+        Write-Host $([Constants]::DoubleDashLine)
+        Write-Host "[Step 4 of 4] Remediating non-compliant Azure IoT Hubs..." 
+        Write-Host $([Constants]::SingleDashLine)
+        
+        if(-not $AutoRemediation)
+        {
+            if (-not $Force)
+            {
+                Write-Host "Found total [$($TLSwithDefaultValue.count)] IoT Hub(s) with default TLS version. TLS version for these resources cannot be reverted back to default value after remediation." -ForegroundColor $([Constants]::MessageType.Warning)
```
Comment (Aboli-msft) [inline Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:443]: cross check on this message

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+        Write-Host $([Constants]::SingleDashLine)
+        Write-Host "Skipped as -DryRun switch is provided." -ForegroundColor $([Constants]::MessageType.Warning)
+        Write-Host $([Constants]::DoubleDashLine)
+
+        Write-Host "Next steps:" -ForegroundColor $([Constants]::MessageType.Info)
+        Write-Host "Run the same command with -FilePath $($backupFile) and without -DryRun, configure TLS Version on IoT Hub(s) listed in the file."
+    }
+}
+
+function Reset-IoTHubTLSVersion
```
Comment (Aboli-msft) [inline Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:595]: Reset-MinTLSVersionForIoTHub rename

Comment (rvadeghar) [reply Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1]: Replaced with suggested command name

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 [hunk 4]
```diff
@@ -0,0 +1,878 @@
...
+        else {
+            throw "API call failed with status code $($response.StatusCode)"
+        }
+    }
+    catch {
+        Write-Error "Error occurred: $_"
+    }
+}
+
+function Set-IoTHubTLSVersion {
```
[the remark this answers is not in the document]
Comment (rvadeghar) [reply Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1]: Replaced with suggested command name

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 (hunk 2 above)
Comment (rvadeghar) [reply Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:443]: Removed that message as it is not needed/applicable on this BRS

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 (hunk 1 above)
Comment (rvadeghar) [reply Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:384]: Updated message as suggested

**Review by Aboli-msft — APPROVED** [review]

File: Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1 [hunk 5]
```diff
@@ -0,0 +1,870 @@
...
+        if(($resourceDetails | Measure-Object).Count -eq 0 -or ($validResources | Measure-Object).Count -eq 0)
+        {
+            Write-Host "No IoT Hub(s) found in input json file for remediation." -ForegroundColor $([Constants]::MessageType.Error)
+            Write-Host $([Constants]::SingleDashLine)
+            return
+        }
+        $validResources | ForEach-Object { 
+            try
+            {
+                $iotHubResourceAPI = "https://management.azure.com/subscriptions/$subId/resourceGroups/$($_.ResourceGroupName)/providers/Microsoft.Devices/IotHubs/$($_.ResourceName)?api-version=2023-06-30"
```
Comment (rajatgoyalMS) [inline Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:256]: didn't we have any PS command instead of calling an API?

Comment (rvadeghar) [reply Scripts/RemediationScripts/Remediate-ConfigureIoTHubTLSVersion.ps1:256]: No, there is powershell command to update AzIotHub. But setting MinTLSVersion property with that command is not possible

**Review by rajatgoyalMS — APPROVED** [review]

**Pull request merged**